auth0/auth0-PHP
# Support for PHP 8

### Description
This branch includes work necessary for the SDK to work with the new release of PHP 8. Most of this PR involves making accommodations for updated dependencies, as the library's own code is compatible with PHP 8.

- These changes require us to bump our minimum supported PHP version from 7.1 to 7.3.
- PHPUnit dependency bumped to 9.3 (from 7) as PHP 8 support will not be getting backported.
- Updated unit tests to support syntax changes in PHPUnit 9. Functionally they are unchanged.
- Introduced a reflection workaround to accommodate PHPUnit 8+ dropping support for accessing invocation stacks natively.
- Guzzle dependency bumped to 7.2 as PHP 8 support will not be getting backported. Drops support for Guzzle 6.
- Overrides our JWT dependency to use a patched repo @ https://github.com/auth0/php-jwt/tree/3.3-php8-compatibility (1)
- Improvements to code styling/formatting and docblock definitions

1. PHP 8 introduces [changes to the OpenSSL functions](https://github.com/php/php-src/blob/php-8.0.0RC4/UPGRADING#L408) which cause the JWT library we depend on to break. The author of the library appears to be working on a major refactoring for a future release (which will require non-trivial effort for us to adapt the SDK to those changes), so I've forked the stable build we currently use and have applied a patch to fix the issues until the upstream package is fixed, and we're able to adapt the SDK to the changes they're making to the library.

~~Note: Our CircleCI tests will fail for this PR as we are testing against PHP versions that this PR drops support for.~~ PHP 7.1 and 7.2 CI tests have been removed to make way for these changes.

### Testing these changes locally
CircleCI is already configured to run our standard tests using PHP 8.0 for this PR; you can review the results in the PR timeline. If you'd like to test these changes locally, you should:

- Install PHP 8.0; configure your local environment to use it
- Fork PR branch
- Run 'composer install' to install dependencies
- Run 'composer test-unit' to ensure 8.0 compatibility
- Run 'composer compat' to ensure 7.3+ compatibility
